internal/orchestrator/updates.go

```diff
@@ -279,16 +279,159 @@ func parseEventUpdateFromKV(ctx context.Context, kve nats.KeyValueEntry,
+// are to be applied to the active-condition KV.
+func filterToReconcile(records []*store.ConditionRecord, updateEvts map[string]*v1types.ConditionUpdateEvent) ([]*rctypes.Condition, []*v1types.ConditionUpdateEvent) {
+	currentCRs := map[string]struct{}{}
+
+	// missing conditions in active-conditions KV to be created
+	creates := []*rctypes.Condition{}
+
+	// in-complete status KV updates to be applied
+	updates := []*v1types.ConditionUpdateEvent{}
+
+	stale := func(createdAt, updatedAt time.Time) bool {
+		return time.Since(createdAt) >= rctypes.StaleThreshold &&
+			time.Since(updatedAt) >= rcontroller.StatusStaleThreshold
+	}
+
+	// updates
+	// - The status KV entry does not exist for the condition listed in active-condition condition
+	// - The first incomplete Condition has exceeded the stale threshold
+	for _, cr := range records {
+		firstCond := cr.Conditions[0]
+		currentCRs[firstCond.Target.String()] = struct{}{}
+
+		_, exists := updateEvts[firstCond.Target.String()]
+		if !exists && stale(firstCond.CreatedAt, firstCond.UpdatedAt) {
+			updates = append(updates, failedUpdateEventFromCondition(firstCond))
+		}
+	}
+
+	//  creates
+	// - The non-final Condition Status entry has no corresponding Active Condition Record.
+	for serverID, updateEvt := range updateEvts {
+		_, exists := currentCRs[serverID]
+		if !exists {
+			creates = append(creates, conditionFromUpdateEvent(updateEvt))
+		}
+	}
+
+	return creates, updates
+}
+
+func (o *Orchestrator) activeConditionsToReconcile(ctx context.Context) ([]*rctypes.Condition, []*v1types.ConditionUpdateEvent) {
+	// list condition records from the active-conditions KV
+	records, err := o.repository.List(ctx)
+	if err != nil {
+		o.logger.WithError(err).Error("condition record lookup error")
+	}
+
+	// returned in-complete status KV updates to reconcile
+	updates := []*v1types.ConditionUpdateEvent{}
+
+	// returned missing conditions from active-conditions KV to reconcile
+	creates := []*rctypes.Condition{}
+
+	// filter condition Kinds and records to be reconciled
+	filteredKinds, filteredRecords := filterIncompleteRecords(records)
+
+	// List Conditions in the Status KV and prepare an update payload
+	// for those to be reconciled.
+	for kind := range filteredKinds {
+		// map of serverIDs to condition updates from the status KV
+		updateEvts := map[string]*v1types.ConditionUpdateEvent{}
+
+		// fetch all conditions statuses for kind
+		statusEntries, err := status.GetAllConditions(kind, o.facility)
+		if err != nil {
+			o.logger.WithError(err).WithField("rctypes.kind", string(kind)).
+				Warn("reconciler error in condition status lookup")
+			continue
+		}
+
+		// parse status entries
+		for _, kve := range statusEntries {
+			evt, err := parseEventUpdateFromKV(ctx, kve, kind)
+			if err != nil {
+				o.logger.WithError(err).WithFields(logrus.Fields{
+					"rctypes.kind": string(kind),
+					"kv.key":       kve.Key(),
+				}).Warn("reconciler skipping malformed update")
+				continue
```

Comment:
We should fail a condition on a garbage update. Otherwise it will linger and accumulate (because the misbehaving controller is still doing things).

Reply:
Nice catch, I'm in agreement, although in this case since all we have is a StatusValue key and nothing more, its not possible to re-construct its equivalent Condition. I left a comment on here https://github.com/metal-toolbox/conditionorc/blob/961699928799ae5d6a1d43de442ed848b008fcc3/internal/orchestrator/updates.go#L399

Reply:
We can't just force a status of `failed`?

Reply:
As discussed we ought to send an otel event or increment some sort of metric to catch this.

Reply:
added span collection in [7e6d17c](https://github.com/metal-toolbox/conditionorc/pull/203/commits/7e6d17c1fab7f89e6d22b2dace24a79b54f59b5b)